README.md
# [![MySQLDumpSecure](https://raw.githubusercontent.com/cytopia/icons/master/64x64/mds.png)](http://mysqldump-secure.org) mysqldump-secure

----
**ATTENTION**

Please upgrade to the latest version in order to be notified about the [OpenSSL SMIME Bug](https://github.com/cytopia/mysqldump-secure/issues/21) in case you are affected.


----
**GIT NOTE:**

WHEN CLONING VIA GIT, MAKE SURE TO ALWAYS CHECK OUT THE LATEST TAG.
THE `MASTER` BRANCH IS ALWAYS UNDER DEVELOPMENT AND THEREFORE UNSTABLE.

----

[Features](https://github.com/cytopia/mysqldump-secure#1-features) |
[Installation](https://github.com/cytopia/mysqldump-secure#2-installation) |
[Configuration](https://github.com/cytopia/mysqldump-secure#3-configuration) |
[Usage](https://github.com/cytopia/mysqldump-secure#4-usage) |
[Documentation](https://github.com/cytopia/mysqldump-secure#5-documentation) |
[Contribution](https://github.com/cytopia/mysqldump-secure#6-contribution) |
[License](https://github.com/cytopia/mysqldump-secure#7-license) |
[Version](https://github.com/cytopia/mysqldump-secure#8-version)

[![Build Status](https://travis-ci.org/cytopia/mysqldump-secure.svg?branch=master)](https://travis-ci.org/cytopia/mysqldump-secure)
[![Latest Stable Version](https://poser.pugx.org/cytopia/mysqldump-secure/v/stable)](https://packagist.org/packages/cytopia/mysqldump-secure) [![Total Downloads](https://poser.pugx.org/cytopia/mysqldump-secure/downloads)](https://packagist.org/packages/cytopia/mysqldump-secure) [![Latest Unstable Version](https://poser.pugx.org/cytopia/mysqldump-secure/v/unstable)](https://packagist.org/packages/cytopia/mysqldump-secure) [![License](https://poser.pugx.org/cytopia/mysqldump-secure/license)](http://opensource.org/licenses/MIT)
[![POSIX](https://img.shields.io/badge/posix-100%25-brightgreen.svg)](https://en.wikipedia.org/?title=POSIX)
[![Type](https://img.shields.io/badge/type-%2Fbin%2Fsh-red.svg)](https://en.wikipedia.org/?title=Bourne_shell)

**[Mysqldump-secure](http://mysqldump-secure.org)** is a POSIX compliant wrapper script for `mysqldump` with many features and very strong security in mind.
It will backup every available database (which is readable by the specified user) as a separate file with the possibility to opt out via blacklisting. Dumped databases can optionally be piped directly to gzip or openssl in order to compress and/or encrypt the backup. Encryption is done before the file is written to disk to avoid possible race conditions (See documentation for more info about security measurements).


Find the project website at [https://mysqldump-secure.org](https://mysqldump-secure.org)
Find the whole post at [https://www.everythingcli.org](https://www.everythingcli.org/index.php/2015/06/13/secure-mysqldump-script-with-encryption-and-compression/)

##### Runs on
[![FreeBSD](https://raw.githubusercontent.com/cytopia/icons/master/64x64/freebsd.png)](https://www.freebsd.org)
[![RedHat](https://raw.githubusercontent.com/cytopia/icons/master/64x64/redhat.png)](https://www.redhat.com)
[![CentOS](https://raw.githubusercontent.com/cytopia/icons/master/64x64/centos.png)](https://www.centos.org)
[![Debian](https://raw.githubusercontent.com/cytopia/icons/master/64x64/debian.png)](https://www.debian.org)
[![ArchLinux](https://raw.githubusercontent.com/cytopia/icons/master/64x64/archlinux.png)](https://www.archlinux.org)
[![Ubuntu](https://raw.githubusercontent.com/cytopia/icons/master/64x64/ubuntu.png)](https://www.ubuntu.com)
[![OSX](https://raw.githubusercontent.com/cytopia/icons/master/64x64/osx.png)](https://www.apple.com/osx)

<sub>(If the script runs on any other system not mentioned here, please drop me a note.)</sub>

##### Rock-stable and well tested

Every push to `mysqldump-secure` triggers `travis-ci` which will run hundreds of all kinds of tests against the new code and stress the tool in every possible way with every possible config.

You can find the tests within the [test](test) folder including an automated setup to get a master-slave server with SSL encryption setup. Thoses tests can also be run locally.

See [https://travis-ci.org/cytopia/mysqldump-secure](https://travis-ci.org/cytopia/mysqldump-secure) for checks in action and [test](test) for a description about the tests.


## 1. Features

### Primary Features

* **Encryption** (hybrid encryption: `RSA` and `AES` via `openssl smime` - no password needed)
* **Compression** (`gzip`, `pigz`, `bzip2`, `pbzip2`, `lzop`, `lzma`, `xz`)
* **Tmpwatch** integration (`tmpwatch` or `tmpreaper`)
* **Transaction-safe** / **Consistent** backups across tables (for DBs with: `InnoDB only`-, `mixed`- and `non-InnoDB`- tables)
* **Conditional mysqldump options** (e.g.: apply `--quick` only on DBs > 200MB)
* **Security** (various built-in checks and precautions)
* **SSL Remote Backups**
* **Nagios** / **Icinga** monitoring integration (via [check_mysqldump-secure](https://github.com/cytopia/check_mysqldump-secure))

### Secondary Features

* Custom mysqldump options
* Master/Slave recognition
* Blacklisting
* Whitelisting
* File logging
* Self validation
* Rock stable (see travis for hundreds of checks)
* 100% POSIX compliant

## 2. Installation

### 2.1 Linux, BSD and OSX

```shell
wget https://github.com/cytopia/mysqldump-secure/archive/0.16.5.tar.gz -O - | tar -xz
cd mysqldump-secure
./configure
make
sudo make install
```

### 2.2 OSX
```shell
brew tap cytopia/tap
brew install mysqldump-secure
```


For more detailed instructions go to the **[Install guidelines](https://github.com/cytopia/mysqldump-secure/blob/master/doc/INSTALL.md)**


## 3. Configuration

There are two separate configuration files:

* [/etc/mysqldump-secure.cnf](etc/mysqldump-secure.cnf)
* [/etc/mysqldump-secure.conf](etc/mysqldump-secure.conf)

The first one is to setup the MySQL credentials and the second one configures the behavior of how to backup the databases.

You do not need to worry about file permissions or directories. The script will auto-check the following options and adjust them as required:

* Logfile exists
* Logfile is writeable
* Auto creation of logfile
* Logging turned off automatically
* Destination dir exists
* Destination dir is writeable
* Auto creation of destination dir
* Required system binaries exist
* MySQL credentials are valid


For more detailed instructions go to the [Setup guidelines](https://github.com/cytopia/mysqldump-secure/blob/master/doc/SETUP.md)

## 4. Usage

### 4.1 Usage
```shell
Usage: mysqldump-secure [--conf] [--cron] [--test] [-v[v]]
       mysqldump-secure --help
       mysqldump-secure --version

When invoked without any arguments, it will start dumping databases as
defined in mysqldump-secure.conf.

  --conf          Pass a different configuration file than the default one.
                  E.g.: --conf=/etc/mysqldump-secure-alt.conf

  --cron          Use for cron run. It will only output errors and warnings
                  and will silence all info, debug and trace output.

  --test          Test requirements and exit.
                  Combine with -v or -vv for more verbose output.

  -v              Show debug output.
                  Specify twice (-vv) to also show trace output.
                  Can be combined with --conf  and --test

  -vv             Show debug and trace output.
                  Can be combined with --conf  and --test

  --help          Show this help screen.

  --version       Show version information.


Exit codes

  0               All fine, no fatals, no errors and no warnings occured.

CHANGELOG.md
Version 0.17 (unreleased)
------------


Version 0.16.5
--------------

- [Fix]     Ensure `PATH` variable is set to survive cron runs


Version 0.16.4
--------------

- [Fix]     `empty` keyword is reserved in MySQL 8+


Version 0.16.3
--------------

- [Fix]     Changed SMIME size threshold from 12MB to 1200MB


Version 0.16.2
--------------

- [Fix]     Warn about SMIME bug if db size > 1200MB (https://github.com/cytopia/mysqldump-secure/issues/21)


Version 0.16.1
--------------

- [Enh]     Code refactoring
- [Enh]     Removed tput coloring (making it work under FreeBSD)
- [Enh]     Removed table list, as it was to slow with db: >1000 tables
- [Fix]     Fixed bug on systems with locales that use commas for floats
- [Fix]     Fixed configure to not stop when target directory does not exist


Version 0.16
------------

- [Fix]     Show pipefail-emulation errors and warnings (now works as expected)
- [Fix]     Using `--defaults-file` instead of `--defaults-extra-file` to make sure no unwanted options are loaded.
- [Fix]     Wrong "finished with errors" on `tmpwatch` readonly files
- [Fix]     Exception Handling: Handle all unbound variables
- [Fix]     Catch error in case of wrong values in all 0/1 variables
- [Fix]     Catch `tmpwatch`/`tmpreaper` errors and show warn/err accordingly
- [Fix]     `tmpreaper`/`tmpwatch` values now accept/validate units (m, h, d)
- [Fix]     Output: stdout did not show all deleted files by tmpwatch/tmpreaper
- [Fix]     Testing: Validate MYSQL_OPTS
- [Enh]     Adding info files for each dump file (optionally).
- [Enh]     Logleves for logging to file (0-3)
- [Enh]     Adding info: Show SSL cyphers in use by server (If SSL is enabled)
- [Enh]     Adding info: Show if Server is master or slave
- [Enh]     Adding info: Show server name and version
- [Enh]     Adding info: Show current connection IP/Port or socket
- [Enh]     Improving Logic: Allow $REQUIRE to overwrite $IGNORE
- [Enh]     Improving Option: Allowing wildcards for ignored databases (e.g: IGNORE="*")
- [Enh]     Adding option: Conditional `--quick` dependent on database size
- [Enh]     Adding option: Consistend DB dumps (InnoDB, mixed, other)
- [Enh]     Adding option: SSL Connection
- [Enh]     Adding argument: --verbose argument
- [Enh]     Testing: Adding lots and lots of travis checks.
- [Enh]     Making /tmp folder settable via config
- [Enh]     Output: New debug level: trace
- [Enh]     Output: Colorized logging
- [Enh]     Output: Consolidate logging
- [Enh]     Script: Use #!/bin/sh -u to check against unset variables
- [Enh]     Removed ${_INSTALL_PREFIX} from config file


Version 0.15
------------

- [Fix]     Exception Handling: Show error when file already exists on disk
- [Fix]     Fixed failed count and total dumped count
- [Fix]     [install] Fixed script install permissions
- [Fix]     [Nagios plugin] Submodule udpated
- [Enh]     [conf] Allow to specify custom config via argument
- [Enh]     Cleaned debug output


Version 0.14.1
--------------

- [Fix]     [install] Create man dir if not exists


Version 0.14
------------

- [Fix]     Do not silently ignore empty databases, show that they were skipped with reason.
- [Enh]     Add man page: man 1 mysqldump-secure
- [Enh]     Improved install routine


Version 0.13
------------

- [Fix]     #17 [sec] chmod is done after dumping to disk (now using umask before)
- [Fix]     #16 allow 0700 chmod formats as well as 700
- [Enh]     Added documentation about security measurements
- [Enh]     Improved code readability of integrity checks
- [Enh]     Auto-fix Nagios Log chmod permissions.
- [Enh]     Add chmod setting for Nagios Log in config.
- [Enh]     Added compression preset: pigz


Version 0.12
------------


mysqldump-secure.sh
#!/bin/sh
#
# @author    Patrick Plocke <patrick@plocke.de>
# @gpg key   0x28BF179F
# @date      2015-07-13
# @licencse  MIT http://opensource.org/licenses/MIT
# @version   v0.3
#
# Script to dump databases one by one
#
# Features:
# ---------------------
# * Encryped output via public/private key (optional)
# * Compressed output via gzip (optional)
# * Database black list (optional)
# * Logfile (optional)
# * Custom mysqldump parameters
# * Dumping time, total time
# * Error handling
#
# Exit Codes
# ---------------------
# * 0: Success
# * 1: Script specific: (writeable directory, config file not found, wrong permissions)
# * 2: Required binary not found
# * 3: MySQL connection error
# * 4: MySQL Database dump error


################################################################################
#
# VARIABLES
#
################################################################################

# Configuration
CONFIG_NAME="mysqldump-secure.conf"
CONFIG_FILE="/etc/${CONFIG_NAME}"


# These command line arguments are considered insecure and can lead
# to compromising your data
MYSQL_EVIL_OPTS="--password -p"

# Do not allow to read any other file than the one specified in
# the configuration.
MYSQL_BAD_OPTS="--defaults-extra-file --defaults-file"



################################################################################
#
# HELPER FUNCTIONS
#
################################################################################

# Output to stdout and to file
output() {
	local MSG="${1}"		# Message to output
	local LOG="${2:-0}"		# Log? 1: yes 0: No (defaults to 0)
	local FILE="${3}"		# Logfile path

	printf "%s\n" "${MSG}"
	[ "${LOG}" = "1" ] && printf "%s %s %s\n" "$(date '+%Y-%m-%d')" "$(date '+%H:%M:%S')" "${MSG}" >> "${FILE}"
	return 0
}
# Inline Output to stdout and to file (no newline)
outputi() {
	local MSG="${1}"		# Message to output
	local LOG="${2:-0}"		# Log? 1: yes 0: No (defaults to 0)
	local FILE="${3}"		# Logfile path

	printf "%s" "${MSG}"
	[ "${LOG}" = "1" ] && printf "%s %s %s" "$(date '+%Y-%m-%d')" "$(date '+%H:%M:%S')" "${MSG}" >> "${FILE}"
	return 0
}
# Output to stdout and to file (no time)
outputn() {
	local MSG="${1}"		# Message to output
	local LOG="${2:-0}"		# Log? 1: yes 0: No (defaults to 0)
	local FILE="${3}"		# Logfile path

	printf "%s\n" "${MSG}"
	[ "${LOG}" = "1" ] && printf "%s\n" "${MSG}" >> "${FILE}"
	return 0
}

# Test if argument is an integer
# @return integer	0: is numer | 1 not a number
isint(){
	printf '%d' "$1" >/dev/null 2>&1 && return 0 || return 1;
}

permission() {
	local file
	local perm

	file="${1}"

	# e.g. 640
	if [ "$(uname)" = "Linux" ]; then
		perm="$(stat --format '%a' ${file})"
	else # Darwin or FreeBSD
		perm="$(stat -f "%OLp" ${file})"
	fi

	echo $perm
	return 0
}




################################################################################
#
# ENTRY POINT: ERROR CHECKING
#
################################################################################


############################################################
# Config FIle
############################################################

if [ ! -f "${CONFIG_FILE}" ]; then
	output "[ERR]  Configuration file not found in ${CONFIG_FILE}"
	output "Aborting"
	exit 1
fi
if [ ! -r "${CONFIG_FILE}" ]; then
	output "[ERR]  Configuration file is not readable in ${CONFIG_FILE}"
	output "Aborting"
	exit 1
fi
if [ "$(permission "${CONFIG_FILE}")" != "400" ]; then
	output "[ERR]  Configuration file ${CONFIG_FILE} has dangerous permissions: $(permission "${CONFIG_FILE}")."
	output "[INFO] Fix it to 400"
	output "Aborting"
	exit 1
fi

# Read config file
. "${CONFIG_FILE}"




############################################################
# Logging Options
############################################################

# Be really strict on checking if we are going to log to file
# or not. Also make sure that the logfile is writeable and
# that no other has read permissions to the file.
if [ -z ${LOG} ]; then
	output '[INFO] $LOG variable is empty or not set in ${CONFIG_FILE}'
	output "[INFO] Logging disabled"
	LOG=0
elif [ "${LOG}" = "1" ]; then
	if [ -z ${LOGFILE} ]; then
		output '[WARN] $LOGFILE variable is empty or not set in ${CONFIG_FILE}'
		output "[WARN] Logging disabled"
		LOG=0
	elif [ ! -f "${LOGFILE}" ]; then
		output "[WARN] Logfile does not exist in ${LOGFILE}"
		outputi "[INFO] Trying to create..."

		if ! touch "${LOGFILE}" > /dev/null 2>&1 ; then
			outputn "Failed"
			output  "[ERR]  Failed to create file ${LOGFILE}"
			output  "[WARN] Logging disabled"
			LOG=0
		else
			outputn "OK"
			output  "[INFO] Created file ${LOGFILE}"
			outputi "[INFO] Trying to chmod..."

			if ! chmod 600 "${LOGFILE}" > /dev/null 2>&1 ; then
				outputn "Failed"
				output  "[ERR]  Failed to chmod 600 ${LOGFILE}"
				output  "[WARN] Logging disabled"
				LOG=0
			else
				outputn "OK"
			fi
		fi
	elif [ ! -w "${LOGFILE}" ]; then
		output "[WARN] Logfile ${LOGFILE} not writeable"
		output "[WARN] Logging disabled"
		LOG=0
	elif [ "$(permission "${LOGFILE}")" != "600" ]; then
		output "[ERR]  Logfile has dangerous permissions: $(permission "${LOGFILE}")"
		output "[INFO] Fix it to 600"
		output "[WARN] Logging disabled"
		LOG=0
	else
		echo "" >> "${LOGFILE}"
		echo "---------------------------------------" >> "${LOGFILE}"
		echo "$(date '+%Y-%m-%d') $(date '+%H:%M:%S') Starting" >> "${LOGFILE}"
		output "[INFO] Logging enabled"
	fi
else
	output "[INFO] Logging not enabled in ${CONFIG_FILE}"
	LOG=0
fi




############################################################
# Destination Directory and Prefix
############################################################

# Check if destination dir exists
if [ -z ${TARGET} ]; then
	output '[ERR]  TARGET variable is empty or not set in ${CONFIG_FILE}' $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 1
elif [ ! -d "${TARGET}" ]; then
	output  "[WARN] Destination dir ${TARGET} does not exist" $LOG "${LOGFILE}"
	outputi "[INFO] Trying to create... " $LOG "${LOGFILE}"
	if ! mkdir -p "${TARGET}" > /dev/null 2>&1 ; then
		outputn "Failed" $LOG "${LOGFILE}"
		output  "Aborting" $LOG "${LOGFILE}"
		exit 1
	else
		outputn "Done" $LOG "${LOGFILE}"
		output "[INFO] Adjusting file permissions on ${TARGET}" $LOG "${LOGFILE}"
		chmod 0700 "${TARGET}"
	fi
fi
# Check if destination dir is writeable
if [ ! -w "${TARGET}" ]; then
	output "[WARN] Destination dir ${TARGET} is not writeable" $LOG "${LOGFILE}"
	outputi "[INFO] Trying to chmod... " $LOG "${LOGFILE}"
	if ! chmod 0700 "${TARGET}" > /dev/null 2>&1 ; then
		outputn "Failed" $LOG "${LOGFILE}"
		output "Aborting" $LOG "${LOGFILE}"
		exit 1
	else
		outputn "Done" $LOG "${LOGFILE}"
	fi
	outputi "[INFO] Trying to chown... " $LOG "${LOGFILE}"
	if ! chown "$(whoami)" "${TARGET}" > /dev/null 2>&1 ; then
		outputn "Failed" $LOG "${LOGFILE}"
		output "Aborting" $LOG "${LOGFILE}"
		exit 1
	else
		outputn "Done" $LOG "${LOGFILE}"
	fi
fi
# Check correct permissions of destination dir
if [ "$(permission "${TARGET}")" != "700" ]; then
	output "[ERR]  Target directory has dangerous permissions: $(permission "${TARGET}")." $LOG "${LOGFILE}"
	output "[INFO] Fix it to 700" $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 1
fi
# Check output Prefix
if [ -z ${PREFIX} ]; then
	output '[INFO] $PREFIX variable is empty not set in ${CONFIG_FILE}'$LOG "${LOGFILE}"
	output "[INFO] Using default 'date-time' prefix" $LOG "${LOGFILE}"
	PREFIX="$(date '+%Y-%m-%d')_$(date '+%H-%M')__"
fi



############################################################
# MySQL
############################################################
if [ -z ${MYSQL_CNF_FILE} ]; then
	output '[ERR]  $MYSQL_CNF_FILE variable is empty or not set in ${CONFIG_FILE}' $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 1
fi
if [ ! -f "${MYSQL_CNF_FILE}" ]; then
	output "[ERR]  MySQL Configuration file not found in ${MYSQL_CNF_FILE}" $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 1
fi
if [ ! -r "${MYSQL_CNF_FILE}" ]; then
	output "[ERR]  MySQL Configuration file is not readable in ${MYSQL_CNF_FILE}" $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 1
fi
if [ "$(permission "${MYSQL_CNF_FILE}")" != "400" ]; then
	output "[ERR]  MySQL Configuration file ${MYSQL_CNF_FILE} has dangerous permissions: $(permission "${MYSQL_CNF_FILE}")." $LOG "${LOGFILE}"
	output "[ERR]  Fix it to 400" $LOG "${LOGFILE}"
	output "[ERR]  Change your database password!" $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 1
fi
if ! command -v mysql > /dev/null 2>&1 ; then
	output "[ERR]  'mysql' not found" $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 2
fi
if ! command -v mysqldump > /dev/null 2>&1 ; then
	output "[ERR]  'mysqldump' not found" $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 2
fi
# Testing MySQL connection
if ! $(which mysql) --defaults-extra-file=${MYSQL_CNF_FILE} -e exit > /dev/null 2>&1 ; then
	output "[ERR]  Cannot connect to mysql database. Check credentials in ${MYSQL_CNF_FILE}" $LOG "${LOGFILE}"
	output "Aborting" $LOG "${LOGFILE}"
	exit 3
fi



############################################################
# Bad MySQL Opts
############################################################
for opt in ${MYSQL_OPTS}; do
	for evil in ${MYSQL_EVIL_OPTS}; do
		if [ "${opt}" = "${evil}" ]; then
			output "[ERR]  Insecure mysqldump option found in MYSQL_OPTS: '${evil}'" $LOG "${LOGFILE}"
			output "Aborting" $LOG "${LOGFILE}"
			exit 3
		fi
	done
	for bad in ${MYSQL_BAD_OPTS}; do
		if [ "${opt}" = "${evil}" ]; then
			output "[ERR]  Disallowed mysqldump option found in MYSQL_OPTS: '${bad}'" $LOG "${LOGFILE}"
			output "Aborting" $LOG "${LOGFILE}"
			exit 3
		fi
	done
done



############################################################
# Compression
############################################################
if [ -z ${COMPRESS} ]; then
	output '[INFO] $COMPRESS variable is empty or not set in ${CONFIG_FILE}' $LOG "${LOGFILE}"
	output "[INFO] Compression disabled" $LOG "${LOGFILE}"
	COMPRESS=0
fi
if [ "${COMPRESS}" = "1" ]; then
	if ! command -v gzip > /dev/null 2>&1 ; then
		output "[WARN] 'gzip' not found" $LOG "${LOGFILE}"
		output "[WARN] Disabling compression" $LOG "${LOGFILE}"
		COMPRESS=0
	fi
else
	output "[INFO] Compression not enabled in ${CONFIG_FILE}" $LOG "${LOGFILE}"
fi



############################################################
# Encryption
############################################################
if [ -z ${ENCRYPT} ]; then
	output "[INFO] \$ENCRYPT variable is empty or not set in ${CONFIG_FILE}" $LOG "${LOGFILE}"
	output "[INFO] Encryption disabled" $LOG "${LOGFILE}"
	ENCRYPT=0
fi
if [ "${ENCRYPT}" = "1" ]; then
	if ! command -v openssl > /dev/null 2>&1 ; then
		output "[ERR]  'openssl' not found" $LOG "${LOGFILE}"
		output "Aborting" $LOG "${LOGFILE}"
		exit 2
	fi
	if [ ! -f "${OPENSSL_PUBKEY_PEM}" ]; then
		output "[ERR]  OpenSSL pubkey not found in ${OPENSSL_PUBKEY_PEM}" $LOG "${LOGFILE}"
		output "Aborting" $LOG "${LOGFILE}"
		exit 2
	fi
	if [ -z ${OPENSSL_ALGO_ARG} ]; then
		output '[WARN] $OPENSSL_ALGO_ARG variable is empty not set in ${CONFIG_FILE}' $LOG "${LOGFILE}"
		output "[INFO] Encryption defaults to -aes256" $LOG "${LOGFILE}"
		OPENSSL_ALGO_ARG="-aes256"
	fi
	# Test openssl Algo
	if ! echo "test" | $(which openssl) smime -encrypt -binary -text -outform DER ${OPENSSL_ALGO_ARG} "${OPENSSL_PUBKEY_PEM}" > /dev/null 2>&1 ; then
		output '[ERR]  openssl encryption test failed. Validate $OPENSSL_ALGO_ARG' $LOG "${LOGFILE}"
		output "Aborting" $LOG "${LOGFILE}"
		exit 2
	fi
else
	output "[INFO] Encryption not enabled in ${CONFIG_FILE}" $LOG "${LOGFILE}"
fi



############################################################
# Deletion
############################################################
if [ -z ${DELETE} ]; then
	output '[INFO] $DELETE variable is empty or not set in ${CONFIG_FILE}' $LOG "${LOGFILE}"
	output "[INFO] Deletion of old files disabled" $LOG "${LOGFILE}"
	DELETE=0
fi
if [ "${DELETE}" = "1"  ]; then
	if [ -z ${DELETE_IF_OLDER} ]; then
		output '[WARN] $DELETE_IF_OLDER variable is empty or not set in ${CONFIG_FILE}' $LOG "${LOGFILE}"
		output "[WARN] Deletion of old files disabled" $LOG "${LOGFILE}"
		DELETE=0
	elif ! isint ${DELETE_IF_OLDER} > /dev/null 2>&1 ; then
		output '[WARN] $DELETE_IF_OLDER variable is not a valid integer' $LOG "${LOGFILE}"
		output "[WARN] Deletion of old files disabled" $LOG "${LOGFILE}"
		DELETE=0
	elif [ ${DELETE_IF_OLDER} -lt 1 ]; then
		output '[WARN] $DELETE_IF_OLDER is smaller than 1 hour' $LOG "${LOGFILE}"
		output "[WARN] Deletion of old files disabled" $LOG "${LOGFILE}"
		DELETE=0
	elif ! command -v tmpwatch > /dev/null 2>&1 ; then
		output "[WARN] 'tmpwatch' not found" $LOG "${LOGFILE}"
		output "[WARN] Deletion of old files disabled" $LOG "${LOGFILE}"
		DELETE=0
	fi
else
	output "[INFO] TMPWATCH deletion not enabled in ${CONFIG_FILE}" $LOG "${LOGFILE}"
fi








################################################################################
#
# ENTRY POINT: MAIN
#
################################################################################

# Binaries
MYSQL="$(which mysql)"
MYSQLDUMP="$(which mysqldump)"
GZIP="$(which gzip)"
OPENSSL="$(which openssl)"
TMPWATCH="$(which tmpwatch)"

ERROR=0


############################################################
# Get all databases
############################################################

# Get a list of all databases
outputi "[INFO] Retrieving list of databases... " $LOG "${LOGFILE}"
DATABASES="$( ${MYSQL} --defaults-extra-file=${MYSQL_CNF_FILE} --batch -e 'show databases;')"
DATABASES="$( echo "${DATABASES}" | sed 1d )"
NUM_DB="$(echo "${DATABASES}" | wc -l | tr -d ' ')"
outputn "${NUM_DB}" $LOG "${LOGFILE}"



############################################################
# Dump databases
############################################################

TOTAL_STARTTIME=$(date +%s)
output "[INFO] Backup directory: ${TARGET}" $LOG "${LOGFILE}"

for db in ${DATABASES}; do

	# Skip specified databases
	skip=0
	for ign_db in ${IGNORE}; do
		if [ "${ign_db}" = "${db}" ]; then
			skip=1
		fi
	done

	if [ ${skip} -eq 0 ]; then

		DB_SIZE="$( ${MYSQL} --defaults-extra-file=${MYSQL_CNF_FILE} --batch \
			-e "SELECT SUM(ROUND(((DATA_LENGTH + INDEX_LENGTH ) / 1024 / 1024),2)) AS Size
				FROM INFORMATION_SCHEMA.TABLES
				WHERE TABLE_SCHEMA = '${db}';")"
		DB_SIZE="$(echo "${DB_SIZE}" | tail -n1)"

		starttime=$(date +%s)
		ext=""	# file extension
		if [ ${COMPRESS} -eq 1 ]; then
			if [ ${ENCRYPT} -eq 1 ]; then
				ext=".sql.gz.pem"
				outputi "Dumping:  ${db} (${DB_SIZE} MB) (compressed) (encrypted) " $LOG "${LOGFILE}"
#				${MYSQLDUMP} --defaults-extra-file=${MYSQL_CNF_FILE} ${MYSQL_OPTS} "${db}" | ${GZIP} -9 | ${OPENSSL} smime -encrypt -binary -text -outform DER ${OPENSSL_ALGO_ARG} -out "${TARGET}/${PREFIX}${db}${ext}" "${OPENSSL_PUBKEY_PEM}"
				# execute with POSIX pipestatus emulation
				exec 4>&1
				error_statuses="`(
					(${MYSQLDUMP} --defaults-extra-file=${MYSQL_CNF_FILE} ${MYSQL_OPTS} "${db}" || echo "0:$?" >&3) |
					(${GZIP} -9 || echo "1:$?" >&3)
					(${OPENSSL} smime -encrypt -binary -text -outform DER ${OPENSSL_ALGO_ARG} -out "${TARGET}/${PREFIX}${db}${ext}" "${OPENSSL_PUBKEY_PEM}" || echo "2:$?" >&3)
				) 3>&1 >&4`"
				exec 4>&-
			else
				ext=".sql.gz"
				outputi "Dumping:  ${db} (${DB_SIZE} MB) (compressed) " $LOG "${LOGFILE}"
#				${MYSQLDUMP} --defaults-extra-file=${MYSQL_CNF_FILE} ${MYSQL_OPTS} "${db}" | ${GZIP} -9 > "${TARGET}/${PREFIX}${db}${ext}"
				# execute with POSIX pipestatus emulation
				exec 4>&1
				error_statuses="`(
					(${MYSQLDUMP} --defaults-extra-file=${MYSQL_CNF_FILE} ${MYSQL_OPTS} "${db}" || echo "0:$?" >&3) |
					(${GZIP} -9 > "${TARGET}/${PREFIX}${db}${ext}" || echo "1:$?" >&3)
				) 3>&1 >&4`"
				exec 4>&-


			fi
		else
			if [ ${ENCRYPT} -eq 1 ]; then
				ext=".sql.pem"
				outputi "Dumping:  ${db} (${DB_SIZE} MB) (encrypted) " $LOG "${LOGFILE}"
#				${MYSQLDUMP} --defaults-extra-file=${MYSQL_CNF_FILE} ${MYSQL_OPTS} "${db}" | ${OPENSSL} smime -encrypt -binary -text -outform DER ${OPENSSL_ALGO_ARG} -out "${TARGET}/${PREFIX}${db}${ext}" "${OPENSSL_PUBKEY_PEM}"
				exec 4>&1
				error_statuses="`(
					(${MYSQLDUMP} --defaults-extra-file=${MYSQL_CNF_FILE} ${MYSQL_OPTS} "${db}" || echo "0:$?" >&3) |
					(${OPENSSL} smime -encrypt -binary -text -outform DER ${OPENSSL_ALGO_ARG} -out "${TARGET}/${PREFIX}${db}${ext}" "${OPENSSL_PUBKEY_PEM}" || echo "1:$?" >&3)
				) 3>&1 >&4`"
				exec 4>&-
			else
				ext=".sql"
				outputi "Dumping:  ${db} (${DB_SIZE} MB) " $LOG "${LOGFILE}"
#				${MYSQLDUMP} --defaults-extra-file=${MYSQL_CNF_FILE} ${MYSQL_OPTS} "${db}" > "${TARGET}/${PREFIX}${db}${ext}"
				exec 4>&1
				error_statuses="`(
					(${MYSQLDUMP} --defaults-extra-file=${MYSQL_CNF_FILE} ${MYSQL_OPTS} "${db}" > "${TARGET}/${PREFIX}${db}${ext}" || echo "0:$?" >&3)
				) 3>&1 >&4`"
				exec 4>&-
			fi
		fi

		# We cannot check against $? as the first if uses a pipe | gzip and so we
		# need to check against the exit code of the first pipe status
		#
		# TODO: $PIPESTATUS is not POSIX conform
		# http://cfaj.ca/shell/cus-faq-2.html
		# check run()
#		if [ ${PIPESTATUS[0]} -ne 0 ]; then
#			outputn "ERROR" $LOG "${LOGFILE}"
#			ERROR=1
#		else
#			endtime=$(date +%s)
#			outputn "$(($endtime - $starttime)) sec" $LOG "${LOGFILE}"
#
#			chmod 600 "${TARGET}/${PREFIX}${db}${ext}"
#		fi

		# No errors in POSIX pipestatus emulation
		if [ -z "$error_statuses" ]; then
			endtime=$(date +%s)
			outputn "$(($endtime - $starttime)) sec" $LOG "${LOGFILE}"
			chmod 600 "${TARGET}/${PREFIX}${db}${ext}"
		else
			outputn "ERROR" $LOG "${LOGFILE}"
			ERROR=1
		fi


	else
		output "Skipping: ${db}" $LOG "${LOGFILE}"
	fi
done
TOTAL_ENDTIME=$(date +%s)

if [ $ERROR -ne 0 ]; then
	output "[ERR]  Some errors occured while dumping" $LOG "${LOGFILE}"
else
	output "[INFO] Dumping finished" $LOG "${LOGFILE}"
	output "[INFO] Took $(($TOTAL_ENDTIME - $TOTAL_STARTTIME)) seconds" $LOG "${LOGFILE}"
fi



############################################################
# Delete old Files
############################################################
if [ ${DELETE} -eq 1 ]; then
	output "[INFO] Deleting files older than ${DELETE_IF_OLDER} hours" $LOG "${LOGFILE}"
	DELETED="$(${TMPWATCH} -m ${DELETE_IF_OLDER} -v "${TARGET}/")"
	if [ $? -ne 0 ]; then
		ERROR=1
	fi
	output "${DELETED}" $LOG "${LOGFILE}"
fi



############################################################
# Exit
############################################################

if [ $ERROR -ne 0 ]; then
	# Send bad exit code
	output "[FAIL] Finished with errors" $LOG "${LOGFILE}"
	exit 4
else
	# Send good exit code
	output "[OK]   Finished successfully" $LOG "${LOGFILE}"
	exit 0
fi
